pkg/resource/deploy/step.go

```diff
@@ -709,6 +709,135 @@ func (s *RefreshStep) Apply(preview bool) (resource.Status, StepCompleteFunc, er
+func (s *ImportStep) Type() tokens.Type                            { return s.new.Type }
+func (s *ImportStep) Provider() string                             { return s.new.Provider }
+func (s *ImportStep) URN() resource.URN                            { return s.new.URN }
+func (s *ImportStep) Old() *resource.State                         { return s.old }
+func (s *ImportStep) New() *resource.State                         { return s.new }
+func (s *ImportStep) Res() *resource.State                         { return s.new }
+func (s *ImportStep) Logical() bool                                { return !s.replacing }
+func (s *ImportStep) Diffs() []resource.PropertyKey                { return s.diffs }
+func (s *ImportStep) DetailedDiff() map[string]plugin.PropertyDiff { return s.detailedDiff }
+
+func (s *ImportStep) Apply(preview bool) (resource.Status, StepCompleteFunc, error) {
+	complete := func() { s.reg.Done(&RegisterResult{State: s.new}) }
+
+	// Read the current state of the resource to import. If the provider does not hand us back any inputs for the
+	// resource, it probably needs to be updated. If the resource does not exist at all, fail the import.
+	prov, err := getProvider(s)
+	if err != nil {
+		return resource.StatusOK, nil, err
+	}
+	read, rst, err := prov.Read(s.new.URN, s.new.ID, nil, nil)
+	if err != nil {
+		if initErr, isInitErr := err.(*plugin.InitError); isInitErr {
+			s.new.InitErrors = initErr.Reasons
+		} else {
+			return rst, nil, err
+		}
+	}
+	if read.Outputs == nil {
+		return rst, nil, errors.Errorf("resource '%v' does not exist", s.new.ID)
+	}
+	if read.Inputs == nil {
+		return resource.StatusOK, nil, errors.Errorf(
+			"provider does not support importing resources; please try updating the '%v' plugin",
+			s.new.URN.Type().Package())
+	}
+	s.new.Outputs = read.Outputs
+
+	// Magic up an old state so the frontend can display a proper diff.
+	s.old = resource.NewState(s.new.Type, s.new.URN, s.new.Custom, false, s.new.ID, read.Inputs, read.Outputs,
+		s.new.Parent, s.new.Protect, false, s.new.Dependencies, s.new.InitErrors, s.new.Provider,
+		s.new.PropertyDependencies, false, nil, nil)
+
+	// Check the user inputs using the provider inputs for defaults.
+	inputs, failures, err := prov.Check(s.new.URN, s.old.Inputs, s.new.Inputs, preview)
+	if err != nil {
+		return rst, nil, err
+	}
+	if issueCheckErrors(s.plan, s.new, s.new.URN, failures) {
+		return rst, nil, errors.New("one or more inputs failed to validate")
+	}
+	s.new.Inputs = inputs
+
+	// Diff the user inputs against the provider inputs. If there are any differences, fail the import.
+	diff, err := diffResource(s.new.URN, s.new.ID, s.old.Inputs, s.old.Outputs, s.new.Inputs, prov, preview)
+	if err != nil {
+		return rst, nil, err
+	}
+	s.diffs, s.detailedDiff = diff.ChangedKeys, diff.DetailedDiff
+
+	if diff.Changes != plugin.DiffNone {
+		const message = "inputs to import do not match the existing resource"
```

Comment:
I feel like this message is not great, but I honestly don't have a better suggestion.

Reply:
Yeah, same.

Reply:
could we print out at least some of things that are different?

Reply:
We do that as part of the usual progress/diff view.

Reply:
Great.  Thanks!